#4839 Configurable local plugins folder feature.

GuruRAM opened the pull request:
<!--
1. Please check out and follow our Contributing Guidelines: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md
2. Do not remove any section of the template. If something is not applicable leave it empty but leave it in the PR
3. Please follow the template, otherwise we'll have to ask you to update it and it will take longer until your PR is merged
-->

## What did you implement:

Closes #4839 Configurable local plugins folder

<!--
Briefly describe the feature if no issue exists for this PR
-->

## How did you implement it:
PluginManager.js was extended with the additional method `parsePluginsObject `which parse the plugins definition and return `localPath` and `modules` (plugins). If path can't be retrieved the default path is returned (`.serverless_plugins`).
PackageService.js was extended and now `getExcludes `method now adds localPath to excludes.
<!--
If this is a nontrivial change please briefly describe your implementation so its easy for us to understand and review your code.
-->

## How can we verify it:
Unit tests were added for both classes.
Can be tested with the following yml file:
`service: localplugin
plugins:
  localPath: './sls_plugins'
  modules:
    - my-plugin-test
provider:
  name: aws
  runtime: nodejs6.10
functions:
  hello:
    handler: handler.hello
`

<!--
Add any applicable config, commands, screenshots or other resources
to make it easy for us to verify this works. The easier you make it for us
to review a PR, the faster we can review and merge it.

Examples:
* serverless.yml - Fully functioning to easily deploy changes
* Screenshots - Showing the difference between your output and the master
* Cloud Configuration - List cloud resources and show that the correct configuration is in place (e.g. AWS CLI commands)
* Other - Anything else that comes to mind to help us evaluate
-->

## Todos:

- [x] Write tests
- [ ] Write documentation - when the logic is settled, documentation will be updated
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO - Because the rollback logic is present

HyperBrain commented on lib/plugins/package/lib/packageService.js:
```diff
@@ -26,9 +26,13 @@ module.exports = {
+    const pluginsLocalPath = this.serverless.service.plugins &&
+      this.serverless.service.plugins.localPath;
+    const localPathExcludes = pluginsLocalPath &&
+      typeof pluginsLocalPath === 'string' ? [pluginsLocalPath] : [];
```
You should use lodash's `_.isString()` here

GuruRAM replied:
Done!

HyperBrain commented on lib/classes/PluginManager.js:
```diff
@@ -130,16 +130,32 @@ class PluginManager {
+
+    if (Array.isArray(servicePlugs)) {
+      modules = servicePlugs;
+    } else if (servicePlugs && Array.isArray(servicePlugs.modules)) {
```
Use lodash's `_.isArray()` here

HyperBrain commented on lib/classes/PluginManager.js:
```diff
@@ -130,16 +130,32 @@ class PluginManager {
+    } else if (servicePlugs && Array.isArray(servicePlugs.modules)) {
+      modules = servicePlugs.modules;
+      localPath = servicePlugs.localPath &&
+        typeof servicePlugs.localPath === 'string' ? servicePlugs.localPath : localPath;
```
`_.isString()`

GuruRAM replied:
I have replaced native calls to lodash isObject, isString and isArray.

HyperBrain commented on lib/classes/PluginManager.test.js:
```diff
@@ -779,6 +779,64 @@ describe('PluginManager', () => {
   });
 
+  describe('#parsePluginsObject()', () => {
+    const parsePluginsObject = (servicePlugins, expectedResult) => {
```
You should name this helper function `parsePluginsObjectAndVerifyResult`. Then it is clear that you do not call the original `parsePluginsObject` function directly in the tests below, but combine verification and call in one function.

GuruRAM replied:
Thank you, I have renamed the name to make it more clear.

GuruRAM commented:
@HyperBrain I have fixed the changes regarding lodash and function name

horike37 commented:
@GuruRAM @HyperBrain 
I tested locally with the plugin command but unfortunately the command is broken by this functionality. As the specification, `sls plugin install` automatically add installed plugins array under `plugins` section in serverless.yml so we can't specify object value there.

I think would be good to move `localPath` under provider section something like this.

```
provider:
  pluginLocalPath: './sls_plugins'

plugins:
  - your-own-local-plugin
  - community-plugin-1
  - community-plugin-2
```

What do you think?

HyperBrain commented:
@horike37 What about moving the plugins declaration to a dual object like we did with the service declaration? So that legacy components like the plugins command can use it too - or adjust the plugins command to use the new semantics.

Moving the `localPath` property outside of the plugins property is imo not a good solution, because it only defines the local path for the plugins, not the provider or anything else.

GuruRAM commented:
@horike37 @HyperBrain Imo the plugins declaration should be in one place and not spread between different sections.

The only thing I'm not comfortable with is the `parsePluginsObject` logic will partially reside in `packageService.js`, `PluginManager.js` and `plugin/install.js` but it can be moved here for example: `plugins\plugin\utils.js` (not sure about the place though).

horike37 commented:
@HyperBrain @GuruRAM 
That makes sense. I think it would be a better approach than mine :+1: 

Currently, adding and removing plugins array are done in the following place.
You can realize the solution by adding if-check whether is an array or object.
https://github.com/serverless/serverless/blob/master/lib/plugins/plugin/install/install.js#L132
https://github.com/serverless/serverless/blob/master/lib/plugins/plugin/uninstall/uninstall.js#L100

GuruRAM commented:
@horike37 @HyperBrain I had to change the implementation of `yamlAstParser`.
The main problem was with branch overriding during add / remove operations.
If the remove plugin2 operation will be performed on the following file:
```
plugins:
  localPath: 'test'
  modules:
    -  plugin1
    -  plugin2
```
It will be transformed to:
```
plugins:
  modules:
    -  plugin1
```
Because the current implementation replaces the the root element (plugins in our case) with the new one and copy only an array into the new root element:
```
    let newObject = JSON.stringify(newArray);
    _.forEach(pathInYmlArrray.reverse(), (v) => {
      newObject = `{"${v}":${newObject}}`;
    });
```
I tried different implementations:

1. I tried to replace only the array element without replacing the whole root element. But it was rather complicated and required some YAML hacks (like manual indent management).
2. **Implemented solution**.  The root element is replaced but the new element contains all the nodes of the old one. I have created `constructPlainObject`  method that constructs configuration plain object from the AST Node.
Unit tests are extended accordingly.

horike37 commented:
@GuruRAM 
You don't have to change codes inside `yamlAstParser` since you can change second node with the current functionality of `yamlAstParser.addNewArrayItem` and `removeExistingArrayItem` something like this

```
return yamlAstParser.removeExistingArrayItem(serverlessFilePath, 'plugins.modules', this.options.pluginName);
```

```
return yamlAstParser.addNewArrayItem(serverlessFilePath, 'plugins.modules', this.options.pluginName);
```
Therefore, You need to check whether to be array or object under `plugins` element and you will apply `plugins.modules` to the second argument of these methods if it is object.

GuruRAM commented:
@horike37 Initially I tried it. The current functionality will remove localPath property from the plugins object. It can be easily tested by extending existing `yamlASTParser.test.js`

Here is the sample test:
```
    it.only('insertion into complex object test', () => {
      const yamlFilePath = path.join(tmpDirPath, 'test.yaml');

      writeFileSync(yamlFilePath, {
        plugins: {
          modules: ['foo'],
          localPath: 'test'
        }
      });
      return expect(yamlAstParser.addNewArrayItem(yamlFilePath, 'plugins.modules', 'bar'))
        .to.be.fulfilled.then(() => {
          const yaml = readFileSync(yamlFilePath, 'utf8');
          expect(yaml).to.be.deep.equal({
            plugins: {
              modules: ['foo', 'bar'],
              localPath: 'test'
            }});
          });
    });
```

The output will be:
```
{
            plugins: {
              modules: ['foo', 'bar']
            }
}
```

And `removeExistingArrayItem` behaves the same way it will remove localPath property since its recreating the whole path from the leaf till the root but only the current leaf will be included into the branch. There is the example:
```
    it.only('deletion from complex object test', () => {
      const yamlFilePath = path.join(tmpDirPath, 'test.yaml');

      writeFileSync(yamlFilePath, {
        plugins: {
          modules: ['foo', 'bar'],
          localPath: 'test'
        }
      });
      return expect(yamlAstParser.removeExistingArrayItem(yamlFilePath, 'plugins.modules', 'bar'))
        .to.be.fulfilled.then(() => {
          const yaml = readFileSync(yamlFilePath, 'utf8');
          expect(yaml).to.be.deep.equal({
            plugins: {
              modules: ['foo'],
              localPath: 'test'
            }});
          });
    });
```

horike37 commented:
@GuruRAM 
Ah, sorry. I missed the case which those is mixed together, so we need to extend the `yamlAstParser` so that the case can be worked as you stated above comment.

GuruRAM commented:
@horike37 no problems at all! I have extended the functionality and unit tests already so these cases are also covered. Can you please check unit tests and the implementation.

HyperBrain commented on lib/classes/PluginManager.js:
```diff
@@ -54,9 +54,9 @@ class PluginManager {
+      if (_.isString(pluginInstance.provider)) {
         pluginProvider = pluginInstance.provider;
-      } else if (typeof pluginInstance.provider === 'object') {
+      } else if (_.isObject(typeof pluginInstance.provider)) {
```
The typeof is superfluous and should be removed: `_.isObject(pluginInstance.provider)`

GuruRAM replied:
@HyperBrain fixed, refactoring type.

HyperBrain commented:
@horike37 Looks good for me now. Can you do a final review regarding the plugin command and the unit tests?

GuruRAM commented:
@horike37 @HyperBrain Cool, thank you!
I have updated the documentation for plugins. I have also updated #installing-plugins to reflect the second available format for the plugins section that is supported.

horike37 commented:
@GuruRAM 
Thank you for the updates! Looks like perfect :smile: 
Currently, we are in merge freeze term for preparation of v1.27.0 as the next version.
This will be shipped in v1.28.0 release!

HyperBrain commented:
@horike37 As you did the important plugin command verifications, could you merge this one in as soon as you are ok with it?

GuruRAM commented:
@horike37 I have merged the latest master to the pull request branch.

horike37 commented:
@GuruRAM 
Thanks for pinging! just merged
